File: solutions/zig/resistor-color-trio/1/resistor_color_trio.zig
const std = @import("std");
const mem = std.mem;

pub const ColorBand = enum(usize) {
    black = 0,
    brown = 1,
    red = 2,
    orange = 3,
    yellow = 4,
    green = 5,
    blue = 6,
    violet = 7,
    grey = 8,
    white = 9,
};

pub fn label(allocator: mem.Allocator, colors: []const ColorBand) mem.Allocator.Error![]u8 {
    const base: u128 = @intFromEnum(colors[0]) * 10 + @intFromEnum(colors[1]);
    const exp = @intFromEnum(colors[2]);
    const ohms: u128 = base * std.math.pow(u128, 10, exp);
    const unit: []const u8 = if (ohms >= 1_000_000_000) "giga"
        else if (ohms >= 1_000_000) "mega"
        else if (ohms >= 1_000) "kilo"
        else "";
    var f_ohms: f64 = @floatFromInt(ohms);
    if (f_ohms >= 1_000_000_000) {
        f_ohms /= 1_000_000_000;
    } else if (ohms >= 1_000_000) {
        f_ohms /= 1_000_000;
    } else if (ohms >= 1_000) {
        f_ohms /= 1_000;
    }

    return std.fmt.allocPrint(allocator, "{d} {s}ohms", .{f_ohms, unit});
}

File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do

  @colors %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  defp get_unit(value) when value >= 1_000_000_000, do: %{divisor: 1_000_000_000.0, prefix: :gigaohms}
  defp get_unit(value) when value >= 1_000_000, do: %{divisor: 1_000_000.0, prefix: :megaohms}
  defp get_unit(value) when value >= 1_000, do: %{divisor: 1_000.0, prefix: :kiloohms}
  defp get_unit(_value), do: %{divisor: 1.0, prefix: :ohms}

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """

  def value(colors), do: colors |> Enum.take(2) |> Enum.map(&Map.get(@colors, &1)) |> Enum.join() |> String.to_integer() |> Kernel.*(10 ** Map.get(@colors, Enum.at(colors, 2)))
  
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors), do: colors |> value() |> then(&{&1 / get_unit(&1).divisor, get_unit(&1).prefix})
end
